site/app/controllers/ChatroomController.php

```diff
@@ -0,0 +1,282 @@
+            $msg_array['host_name'] = $chatroom->getHostName();
+            $msg_array['base_url'] = $this->core->buildCourseUrl(['chat']);
+            $msg_array['socket'] = "chatrooms";
+        }
+        else {
+            $msg_array = [];
+            $msg_array['type'] = 'chat_close';
+            $msg_array['id'] = $chatroom->getId();
+            $msg_array['socket'] = "chatrooms";
+            $indiv_msg_array = [];
+            $indiv_msg_array['type'] = 'chat_close';
+            $indiv_msg_array['socket'] = "chatroom_$chatroom_id";
+            $this->sendSocketMessage($indiv_msg_array);
+        }
+        $this->sendSocketMessage($msg_array);
+        $chatroom->setSessionStartedAt($chatroom->isActive() ? null : new \DateTime("now"));
+        $chatroom->toggleActiveStatus();
+
+        $em->flush();
+
+        return new RedirectResponse($this->core->buildCourseUrl(['chat']));
+    }
+
+    #[Route("/courses/{_semester}/{_course}/chat/{chatroom_id}/messages", methods: ["GET"], requirements: ["chatroom_id" => "\d+"])]
+    public function fetchMessages(string $chatroom_id): JsonResponse {
+        $em = $this->core->getCourseEntityManager();
+        $messages = $em->getRepository(Message::class)->findBy(['chatroom' => $chatroom_id], ['timestamp' => 'ASC']);
+
+        $formattedMessages = array_map(function ($message) {
+            return [
+                'id' => $message->getId(),
+                'content' => $message->getContent(),
+                'timestamp' => $message->getTimestamp()->format('Y-m-d H:i:s'),
+                'user_id' => $message->getUserId(),
+                'display_name' => $message->getDisplayName(),
+                'role' => $message->getRole()
+            ];
+        }, $messages);
+
+        return JsonResponse::getSuccessResponse($formattedMessages);
+    }
+
+    #[Route("/courses/{_semester}/{_course}/chat/{chatroom_id}/send/{anonymous_route_segment?}", methods: ["POST"], requirements: ["chatroom_id" => "\d+", "anonymous_route_segment" => "anonymous"])]
+    public function addMessage(string $chatroom_id, ?string $anonymous_route_segment = null,): JsonResponse {
+        $isAnonymous = ($anonymous_route_segment === 'anonymous');
+        $em = $this->core->getCourseEntityManager();
+        $user = $this->core->getUser();
+        $chatroom = $em->getRepository(Chatroom::class)->find($chatroom_id);
+        if ($chatroom === null) {
+            return JsonResponse::getFailResponse("Chatroom not found");
+        }
+        if (!$chatroom->isActive() && !$user->accessAdmin()) {
+            return JsonResponse::getFailResponse("This chatroom is not enabled");
+        }
+        if (strcmp($_POST['content'], "") === 0) {
+            return JsonResponse::getFailResponse("Can't send blank message");
+        }
+        $msg_array = [];
+        $msg_array['content'] = $_POST['content'];
+        $msg_array['type'] = 'chat_message';
+        $msg_array['user_id'] = $user->getId();
```

Comment:
If you are going to send `user_id`, that removes all form of anonymity as a user could just open up chrome dev tools and see the user id.